acra-censor/common/logging_logic.go

```diff
@@ -0,0 +1,268 @@
+}
+
+// NewFileQueryWriter creates QueryWriter instance
+func NewFileQueryWriter(filePath string) (*QueryWriter, error) {
+	// create writer
+	writer := &QueryWriter{}
+	storage, err := NewFileLogStorage(filePath)
+	if err != nil {
+		writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't create QueryWriter instance")
+	}
+	writer.logStorage = storage
+	writer.queryIndex = 0
+	writer.mutex = &sync.Mutex{}
+	// signals
+	signalShutdown := make(chan os.Signal, 2)
+	signal.Notify(signalShutdown, os.Interrupt, syscall.SIGTERM)
+	signalBackgroundExit := make(chan bool)
+	writer.signalBackgroundExit = signalBackgroundExit
+
+	writer.serializationTimeout = DefaultSerializationTimeout
+	writer.serializationTicker = time.NewTicker(DefaultSerializationTimeout)
+	writer.logger = log.WithField("internal_object", "querywriter")
+
+	// read existing queries
+	err = writer.readStoredQueries()
+	if err != nil {
+		writer.logger.WithError(ErrCantReadQueriesFromFileError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't create QueryWriter instance")
+		writer.logStorage.Close()
+		return nil, err
+	}
+
+	//handling goroutine
+	go func() {
+		for {
+			select {
+			case <-writer.serializationTicker.C:
+				err := writer.dumpBufferedQueries()
+				if err != nil {
+					writer.logger.WithError(ErrComplexSerializationError).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Can't dump buffered queries")
+				}
+				writer.serializationTicker.Stop()
+				writer.serializationTicker = time.NewTicker(writer.serializationTimeout)
+			case <-signalBackgroundExit:
+				writer.serializationTicker.Stop()
+				err := writer.logStorage.Close()
+				if err != nil {
+					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Error occurred on exit QueryWriter instance")
+				}
+			case <-signalShutdown:
+				writer.serializationTicker.Stop()
+				err := writer.logStorage.Close()
+				if err != nil {
+					writer.logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorIOError).Errorln("Error occurred on shutdown QueryWriter instance")
+				}
+				return
+			default:
+				//do nothing. This means that channel has no data to read yet
+			}
+		}
+	}()
+	return writer, nil
+}
+
+// SetSerializationTimeout sets timeout of dumping captured queries to the file.
+func (queryWriter *QueryWriter) SetSerializationTimeout(timeout time.Duration) {
+	queryWriter.serializationTimeout = timeout
+}
+
+// GetSerializationTimeout gets timeout of dumping captured queries to the file.
+func (queryWriter *QueryWriter) GetSerializationTimeout() time.Duration {
+	return queryWriter.serializationTimeout
+}
+
+// DumpQueries writes all queries into file.
+func (queryWriter *QueryWriter) DumpQueries() error {
+	rawData := queryWriter.serializeQueries(queryWriter.Queries)
```

Comment:
here we need lock to avoid background append to Queries

Reply:
after serialization, we can unlock and avoid lock on writing to file because we have serialized copy of data